### PR TITLE
Added a flag to fetch simple balance vs confirmed / pending balance

### PR DESCRIPTION
**Pull request by drcpu-github**
Fetching confirmed and pending balance is terribly slow for executing in an automated fashion (e.g. as part of a block explorer or a simple wallet app). I added a flag to the balance RPC call to fall back to the simple balance counter instead of the confirmed / pending balance.

### CODE REVIEW

**Comment by tmpolaczyk** [comment]
So there are two reasons why getBalance is so slow:

* The balance is not indexed, which means that to calculate it we need to iterate over all the UTXOs and sum the ones that belong to the provided PublicKeyHash.
* The distinction of confirmed and unconfirmed balance means that we need to check the UTXOs that existed before the last superblock, and currently this is implemented by reading the old ChainState from storage and iterating again. This will probably be fixed by implementing #1583 or something similar.

This PR fixes the second point, and it's a good idea because most of the time the confirmed balance will be the same as the unconfirmed.

File: node/src/actors/json_rpc/json_rpc_methods.rs
Comment (tmpolaczyk) [inline node/src/actors/json_rpc/json_rpc_methods.rs:1114]: Changing the input type is technically a breaking change, but it may not be a problem because getBalance is only used in the CLI, and we may break the JSON-RPC API in PR #1939 anyway.

But if we want to make this backwards-compatible, this method should take an unparsed `Params`, similar to the `get_block` method, and it should parse the input manually: if params is an array, parse using old logic and set simple to false; if params is an object parse using new logic.

```suggestion
    params: Params,
```

File: data_structures/src/transaction_factory.rs
Comment (tmpolaczyk) [inline data_structures/src/transaction_factory.rs:269]: Instead of duplicating the total balance, the `confirmed` field should be an `Option<u64>`, set to `None` if the confirmed balance is not known.

```suggestion
            confirmed: None,
```

**Comment by drcpu-github** [comment]
> * The balance is not indexed, which means that to calculate it we need to iterate over all the UTXOs and sum the ones that belong to the provided PublicKeyHash.
> * The distinction of confirmed and unconfirmed balance means that we need to check the UTXOs that existed before the last superblock, and currently this is implemented by reading the old ChainState from storage and iterating again. This will probably be fixed by implementing #1583 or something similar.

I actually 'analyzed' where the slowdown is coming from and an overwhelming part comes from fetching the ChainState from storage. Iterating over the UTXOs may not be optimal, but it is nowhere near slow. After the confirmed balance change, fetching balance takes >2s (which is obviously infeasibly slow for tool-based fetching), before the change it took less than 0.1s (on average) if I remember correctly. I'm not quite sure if a change like storing UTXOs in storage will actually help with the speed of a `get_balance` call since it's still fetching from storage, but maybe it does.

I added all of your suggested changes, so when the tests pass, I'll squash them and this PR can be merged.

File: node/src/actors/json_rpc/json_rpc_methods.rs
Comment (tmpolaczyk) [inline node/src/actors/json_rpc/json_rpc_methods.rs:1139]: I didn't mean to copy get_block exactly, this can be removed

```suggestion
        simple = false;
```

File: node/src/actors/json_rpc/json_rpc_methods.rs
Comment (tmpolaczyk) [inline node/src/actors/json_rpc/json_rpc_methods.rs:1143]: And here we should parse params as `GetBalanceParams`, otherwise I think in this state the CLI method doesn't work:

```suggestion
        let params: GetBalanceParams = params.parse()?;
        pkh = params.pkh;
        simple = params.simple;
```

**Comment by tmpolaczyk** [comment]
> I'm not quite sure if a change like storing UTXOs in storage will actually help with the speed of a get_balance call since it's still fetching from storage, but maybe it does.

That's a good point, because iterating over something that is in storage should be slower than iterating over something that is in RAM, but I remember testing it and RocksDB is surprisingly fast. The current slowdown is mostly from deserializing the entire ChainState struct when only the `utxo_set` is needed, and that will be skipped if the UTXOs are stored in separate keys.

File: node/src/actors/json_rpc/json_rpc_methods.rs
Comment (tmpolaczyk) [inline node/src/actors/json_rpc/json_rpc_methods.rs:1105]: This field should not use `#[serde(default)]` because it is a required argument.

```suggestion
```

File: node/src/actors/json_rpc/json_rpc_methods.rs
Comment (drcpu-github) [reply node/src/actors/json_rpc/json_rpc_methods.rs:1139]: I completely misunderstood how that block of code worked, my Rust knowledge is sometimes a bit lacking.

**Comment by drcpu-github** [comment]
```
./witnet -c .witnet/config/escrow.toml node raw

{"jsonrpc": "2.0", "method": "getBalance", "params": ["..."], "id": 1}
{"jsonrpc":"2.0","result":{"confirmed":1966529170,"total":1966529170},"id":1}

{"jsonrpc": "2.0", "method": "getBalance", "params": {"pkh": "...", "simple": true}, "id": 1}
{"jsonrpc":"2.0","result":{"confirmed":null,"total":1966529170},"id":1}

{"jsonrpc": "2.0", "method": "getBalance", "params": {"pkh": "...", "simple": false}, "id": 1}
{"jsonrpc":"2.0","result":{"confirmed":1966529170,"total":1966529170},"id":1}

./witnet -c .witnet/config/witnet.toml node balance

Confirmed balance:   1.966529170 wits
Pending balance:     0.000000000 wits

./witnet -c .witnet/config/witnet.toml node balance --simple

Balance:   1.966529170 wits
```
All of these raw JSON RPC calls + CLI produce the results I expected, so I think this implementation is good. Thanks for the review!

**Comment by drcpu-github** [comment]
> That's a good point, because iterating over something that is in storage should be slower than iterating over something that is in RAM, but I remember testing it and RocksDB is surprisingly fast. The current slowdown is mostly from deserializing the entire ChainState struct when only the `utxo_set` is needed, and that will be skipped if the UTXOs are stored in separate keys.

Yeah, that's the part I'm not too sure about. If the `utxo_set` is a small part of the ChainState, it may result in a noticeably faster completion compared to now, but I would still expect it'll be significantly slower as that data structure is probably quite big. I think this PR is a nice addition overall since it allows for specialization and optimization of balance calls when confirmed vs pending is not required.

**Review by tmpolaczyk — APPROVED** [review]

**Comment by tmpolaczyk** [comment]
Thanks for the PR! Don't forget to squash the commits before merging.

**Comment by drcpu-github** [comment]
I squashed the commits, but I cannot merge it myself as far as I know.

**Pull request merged**
